자바스크립트는 비동기 처리를 위한 하나의 패턴으로 콜백 함수를 사용한다. 하지만 전통적인 콜백 패턴은 콜백 헬로 인해 가독성이 나쁘고 비동기 처리 중 발생한 에러의 처리가 곤란하며 여러 개의 비동기 처리를 한 번에 처리하는 데도 한계가 있다.

ES6에서는 비동기 처리를 위한 또 다른 패턴으로 프로미스를 도입했다. 프로미스는 전통적인 콜백 패턴이 가진 단점을 보완하며 비동기 처리 시점을 명확하게 표현할 수 있다는 장점이 있다.

## 45.1 비동기 처리를 위한 콜백 패턴의 단점

#### 45.1.1 콜백 헬

먼저 GET 요청을 위한 함수를 작성해 보자.

```javascript
// GET 요청을 위한 비동기 함수
const get = url => {
  const xhr = new XMLHttpRequest();
  xhr.open('GET', url);
  xhr.send();

  xhr.onload = () => {
    if (xhr.status === 200) {
      // 서버의 응답을 콘솔에 출력한다.
      console.log(JSON.parse(xhr.response));
    } else {
      console.error(`${xhr.status} ${xhr.statusText}`);
    }
  };
};

// id가 1인 post를 취득
get('https://jsonplaceholder.typicode.com/posts/1');
/*
{
  "userId": 1,
  "id": 1,
  "title": "sunt aut facere ...",
  "body": "quia et suscipit ..."
}
*/
```

위 예제의 `get` 함수는 서버의 응답 결과를 콘솔에 출력한다. `get` 함수가 서버의 응답 결과를 반환하게 하려면 어떻게 하면 될까?

`get` 함수는 비동기 함수다. 비동기 함수란 내부에 비동기로 동작하는 코드를 포함한 함수를 말한다. **비동기 함수를 호출하면 함수 내부의 비동기로 동작하는 코드가 완료되지 않았다 해도 기다리지 않고 즉시 종료된다. 즉, 비동기 함수 내부의 비동기로 동작하는 코드는 비동기 함수가 종료된 이후에 완료된다. 따라서 비동기 함수 내부의 비동기로 동작하는 코드에서 처리 결과를 외부로 반환하거나 상위 스코프의 변수에 할당하면 기대한 대로 동작하지 않는다.**

예를 들어, `setTimeout` 함수는 비동기 함수다. `setTimeout` 함수가 비동기 함수인 이유는 콜백 함수의 호출이 비동기로 동작하기 때문이다. `setTimeout` 함수를 호출하면 호출 스케줄링한 다음, 타이머 `id` 를 반환하고 즉시 종료된다. 즉, 비동기 함수인 `setTimeout` 함수의 콜백 함수는 `setTimeout` 함수가 종료된 이후에 호출된다. 따라서 `setTimeout` 함수 내부의 콜백 함수에서 처리 결과를 외부로 반환하거나 상위 스코프의 변수에 할당하면 기대한 대로 동작하지 않는다.

`setTimeout` 함수의 콜백 함수에서 상위 스코프의 변수에 값을 할당해 보자. `setTimeout` 함수는 생성된 타이머를 식별할 수 있는 고유한 타이머 `id` 를 반환하므로 콜백 함수에서 값을 반환하는 것은 무의미하다.

```javascript
let g = 0;

// 비동기 함수인 setTimeout 함수는 콜백 함수의 처리 결과를 외부로 반환하거나
// 상위 스코프의 변수에 할당하지 못한다.
setTimeout(() => { g = 100; }, 0);
console.log(g); // 0
```

GET 요청을 전송하고 서버의 응답을 전달받는 `get` 함수도 비동기 함수다. `get` 함수가 비동기 함수인 이유는 `get` 함수 내부의 **`onload` 이벤트 핸들러가 비동기로 동작**하기 때문이다. `get` 함수를 호출하면 GET 요청을 전송하고 `onload` 이벤트 핸들러를 등록한 다음 `undefined` 를 반환하고 즉시 종료된다. 즉, 비동기 함수인 `get` 함수 내부의 `onload` 이벤트 핸들러는 `get` 함수가 종료된 이후에 실행된다. 따라서 `get` 함수의 `onload` 이벤트 핸들러에서 서버의 응답 결과를 반환하거나 상위 스코프의 변수에 할당하면 기대한 대로 동작하지 않는다.

`get` 함수가 서버의 응답 결과를 반환하도록 수정해 보자.

```javascript
// GET 요청을 위한 비동기 함수
const get = url => {
  const xhr = new XMLHttpRequest();
  xhr.open('GET', url);
  xhr.send();

  xhr.onload = () => {
    if (xhr.status === 200) {
      // ① 서버의 응답을 반환한다.
      return JSON.parse(xhr.response);
    }
    console.error(`${xhr.status} ${xhr.statusText}`);
  };
};

// ② id가 1인 post를 취득
const response = get('https://jsonplaceholder.typicode.com/posts/1');
console.log(response); // undefined
```

`get` 함수가 호출되면 `XMLHttpRequest` 객체를 생성하고, HTTP 요청을 초기화한 후, HTTP 요청을 전송한다. 그리고 `xhr.onload` 이벤트 핸들러 프로퍼티에 이벤트 핸들러를 바인딩하고 종료한다. 이때 `get` 함수에 명시적인 반환문이 없으므로 `get` 함수는 `undefined` 를 반환한다(2).

`xhr.onload` 이벤트 핸들러 프로퍼티에 바인딩한 이벤트 핸들러의 반환문(1)은 `get` 함수의 반환문이 아니다. `get` 함수는 반환문이 생략되었으므로 암묵적으로 `undefined` 를 반환한다. 함수의 반환값은 명시적으로 호출한 다음에 캐치할 수 있으므로 `onload` 이벤트 핸들러를 `get` 함수가 호출할 수 있다면 이벤트 핸들러의 반환값을 `get` 함수가 캐치하여 다시 반환할 수도 있겠지만 `onload` 이벤트 핸들러는 `get` 함수가 호출하지 않기 때문에 그럴 수도 없다. 따라서 `onload` 이벤트 핸들러의 반환값은 캐치할 수 없다.

```html
<!DOCTYPE html>
<html>
<body>
  <input type="text">
  <script>
    document.querySelector('input').oninput = function () {
      console.log(this.value);
      // 이벤트 핸들러에서의 반환은 의미가 없다.
      return this.value;
    };
  </script>
</body>
</html>
```

그렇다면 (1)에서 서버의 응답을 상위 스코프의 변수에 할당하면 어떨까?

```javascript
let todos;

// GET 요청을 위한 비동기 함수
const get = url => {
  const xhr = new XMLHttpRequest();
  xhr.open('GET', url);
  xhr.send();

  xhr.onload = () => {
    if (xhr.status === 200) {
      // ① 서버의 응답을 상위 스코프의 변수에 할당한다.
      todos = JSON.parse(xhr.response);
    } else {
      console.error(`${xhr.status} ${xhr.statusText}`);
    }
  };
};

// id가 1인 post를 취득
get('https://jsonplaceholder.typicode.com/posts/1');
console.log(todos); // ② undefined
```

이 또한 기대한 대로 동작하지 않는다. `xhr.onload` 이벤트 핸들러 프로퍼티에 바인딩한 이벤트 핸들러는 언제나 (2)의 `console.log` 가 종료한 이후에 호출된다. 따라서 (2)의 시점에서는 아직 전역 변수 `todos` 에 서버의 응답 결과가 할당되기 이전이다. 다시 말해, `xhr.onload` 이벤트 핸들러에서 서버의 응답을 상위 스코프의 변수에 할당(1)하면 처리 순서가 보장되지 않는다. 그 이유를 알아보자.

비동기 항수 `get` 이 호출되면 함수 코드를 평가하는 과정에서 `get` 함수의 실행 컨텍스트가 생성되고 실행 컨텍스트 스택(콜 스택)에 푸시된다. 이후 함수 코드 실행 과정에서 `xhr.onload` 이벤트 핸들러 프로퍼티에 이벤트 핸들러가 바인딩된다.

`get` 함수가 종료하면 `get` 함수의 실행 컨텍스트가 콜 스택에서 팝되고, 곧바로 (2)의 `console.log` 가 호출된다. 이때 `console.log` 의 실행 컨텍스트가 생성되어 실행 컨텍스트 스택에 푸시된다. 만약 `console.log` 가 호출되기 직전에 `load` 이벤트가 발생했다 하더라도 `xhr.onload` 이벤트 핸들러 프로퍼티에 바인딩한 이벤트 핸들러는 결코 `console.log` 보다 먼저 실행되지 않는다. 

서버로부터 응답이 도착하면 `xhr` 객체에서 `load` 이벤트가 발생한다. 이때 **`xhr.onload` 핸들러 프로퍼티에 바인딩한 이벤트 핸들러가 즉시 실행되는 것이 아니다. `xhr.onload` 이벤트 핸들러는 `load` 이벤트가 발생하면 일단 태스크 큐에 저장되어 대기하다가, 콜 스택이 비면 이벤트 루프에 의해 콜 스택으로 푸시되어 실행된다.** 이벤트 핸들러도 함수이므로 이벤트 핸들러의 평가 -> 이벤트 핸들러의 실행 컨텍스트 생성 -> 콜 스택에 푸시 -> 이벤트 핸들러 실행 과정을 거친다.

따라서 `xhr.onload` 이벤트 핸들러가 실행되는 시점에는 콜 스택이 빈 상태여야 하므로 (2)의 `console.log` 는 이미 종료된 이후다. 만약 `get` 함수 이후에 `console.log` 가 100번 호출된다 해도 `xhr.onload` 이벤트 핸들러는 모든 `console.log` 가 종료된 이후에 실행된다. 즉, `xhr.onload` 이벤트 핸들러에서 상위 스코프의 변수에 서버의 응답 결과를 할당하기 이전에 `console.log` 가 먼저 호출되어 `undefined` 가 출력된다.

**이처럼 비동기 함수는 비동기 처리 결과를 외부에 반환할 수 없고, 상위 스코프의 변수에 할당할 수도 없다. 따라서 비동기 함수의 처리 결과(서버의 응답 등)에 대한 후속 처리는 비동기 함수 내부에서 수행해야 한다. 이때 비동기 함수를 범용적으로 사용하기 위해 비동기 함수에 비동기 처리 결과에 대한 후속 처리를 수행하는 콜백 함수를 전달하는 것이 일반적이다. 필요에 따라 비동기 처리가 성공하면 호출될 콜백 함수와 비동기 처리가 실패하면 호출될 콜백 함수를 전달할 수 있다.**

```javascript
// GET 요청을 위한 비동기 함수
const get = (url, successCallback, failureCallback) => {
  const xhr = new XMLHttpRequest();
  xhr.open('GET', url);
  xhr.send();

  xhr.onload = () => {
    if (xhr.status === 200) {
      // 서버의 응답을 콜백 함수에 인수로 전달하면서 호출하여 응답에 대한 후속 처리를 한다.
      successCallback(JSON.parse(xhr.response));
    } else {
      // 에러 정보를 콜백 함수에 인수로 전달하면서 호출하여 에러 처리를 한다.
      failureCallback(xhr.status);
    }
  };
};

// id가 1인 post를 취득
// 서버의 응답에 대한 후속 처리를 위한 콜백 함수를 비동기 함수인 get에 전달해야 한다.
get('https://jsonplaceholder.typicode.com/posts/1', console.log, console.error);
/*
{
  "userId": 1,
  "id": 1,
  "title": "sunt aut facere ...",
  "body": "quia et suscipit ..."
}
*/
```

이처럼 콜백 함수를 통해 비동기 처리 결과에 대한 후속 처리를 수행하는 비동기 함수가 비동기 처리 결과를 가지고 또다기 비동기 함수를 호출해야 한다면 콜백 함수 호출이 중첩되어 복잡도가 높아지는 현상이 발생하는데, 이를 콜백 헬<sup>callback hell</sup> 이라 한다.

```javascript
// GET 요청을 위한 비동기 함수
const get = (url, callback) => {
  const xhr = new XMLHttpRequest();
  xhr.open('GET', url);
  xhr.send();

  xhr.onload = () => {
    if (xhr.status === 200) {
      // 서버의 응답을 콜백 함수에 전달하면서 호출하여 응답에 대한 후속 처리를 한다.
      callback(JSON.parse(xhr.response));
    } else {
      console.error(`${xhr.status} ${xhr.statusText}`);
    }
  };
};

const url = 'https://jsonplaceholder.typicode.com';

// id가 1인 post의 userId를 취득
get(`${url}/posts/1`, ({ userId }) => {
  console.log(userId); // 1
  // post의 userId를 사용하여 user 정보를 취득
  get(`${url}/users/${userId}`, userInfo => {
    console.log(userInfo); // {id: 1, name: "Leanne Graham", username: "Bret",...}
  });
});
```

위 예제를 보면 GET 요청을 통해 서버로부터 응답(`id` 가 1인 `post`)을 취득하고 이 데이터를 사용하여 또다시 GET 요청을 한다. 콜백 헬은 가독성을 나쁘게 하며 실수를 유발하는 원인이 된다. 

```javascript
get('/step1', a => {
  get(`/step2/${a}`, b => {
    get(`/step3/${b}`, c => {
      get(`/step4/${c}`, d => {
        console.log(d);
      });
    });
  });
});
```

#### 45.1.2 에러 처리의 한계

비동기 처리를 위한 콜백 패턴의 문제점 중에서 가장 심각한 것은 에러 처리가 곤란하다는 것이다.

```javascript
try {
  setTimeout(() => { throw new Error('Error!'); }, 1000);
} catch (e) {
  // 에러를 캐치하지 못한다
  console.error('캐치한 에러', e);
}
```

`try` 코드 블록 내에서 호출한 `setTimeout` 함수는 1초 후에 콜백 함수가 실행되도록 타이머를 설정하고, 이후 콜백 함수는 에러를 발생시킨다. 하지만 이 에러는 `catch` 코드 블록에서 캐치되지 않는다. 

비동기 함수인 `setTimeout` 이 호출되면 `setTimeout` 함수의 실행 컨텍스트가 생성되어 콜 스택에 푸시되어 실행된다. `setTimeout` 는 비동기 함수이므로 콜백 함수가 호출되는 것을 기다리지 않고 즉시 종료되어 콜 스택에서 제거된다. 이후 타이머가 만료되면 `setTimeout` 함수의 콜백 함수는 태스크 큐로 푸시되고 콜 스택이 비어졌을 때 이벤트 루프에 의해 콜 스택으로 푸시되어 실행된다.

`setTimeout` 함수의 콜백 함수가 실행될 때 `setTimeout` 함수는 이미 콜 스택에서 제거된 상태다. 이것은 `setTimeout` 함수의 콜백 함수를 호출한 것이 `setTimeout` 함수가 아니라는 것을 의미한다. `setTimeout` 함수의 콜백 함수의 호출자가 `setTimeout` 함수라면 콜 스택의 현재 실행 중인 실행 컨텍스트가 콜백 함수의 실행 컨텍스트일 때 현재 실행 중인 실행 컨텍스트의 하위 실행 컨텍스트가 `setTimeout` 함수이어야 한다. 

**에러르 호출자 방향으로 전파된다.** 즉, 콜 스택의 아래 방향(실행 중인 실행 컨텍스트가 푸시되기 직전에 푸시된 실행 컨텍스트 방향)으로 전파된다. 하지만 앞에서 살펴본 바와 같이 `setTimeout` 함수의 콜백 함수를 호출한 것은 `setTimeout` 함수가 아니다. 따라서 `setTimeout` 함수의 콜백 함수가 발생시킨 에러는 `catch` 블록에서 캐치되지 않는다.

지금까지 살펴본 비동기 처리를 위한 콜백 패턴은 콜백 헬이나 에러 처리가 곤란하다는 문제가 있다. 이를 극복하기 위해 ES6에서 프로미스가 도입되었다.

